routing/rt/graph/graph.go

```diff
@@ -0,0 +1,536 @@
+// Distributed under the MIT software license, see the accompanying
+// file LICENSE or http://www.opensource.org/licenses/mit-license.php
+
+package graph
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"math/rand"
+	"os"
+	"strconv"
+	"bytes"
+)
+
+var (
+	CyclicDataError         = errors.New("Cyclic data")
+	PathNotFoundError       = errors.New("Path not found")
+	NodeNotFoundError       = errors.New("Node not found")
+	EdgeNotFoundError       = errors.New("Edge not found")
+)
+
+type ID struct {
+	ID [33]byte
+}
+
+var NilID = ID{ID: [33]byte{}}
+
+func NewIDDebug(s interface{}) ID {
+	switch id := s.(type) {
+	case string:
+		x := ID{}
+		copy(x.ID[:], []byte(id))
+		return x
+	case []byte:
+		x := ID{}
+		copy(x.ID[:], id)
+		return x
+	case [32]byte:
+		return NewIDDebug(id[:])
+	case int:
+		return NewIDDebug(strconv.Itoa(id))
+	default:
+		return NewIDDebug(fmt.Sprintf("%v", id))
+	}
+}
+
+func NewID(s [33]byte) ID{
+	return ID{ID:s}
+}
+
+func (s ID) String() string {
+	return string(s.ID[:])
+}
+
+func (s ID) ToByte32() [32]byte {
+	var rez [32]byte
+	copy(rez[:], s.ID[:])
+	return rez
+}
+
+func (s ID) ToByte33() [33]byte {
+	var rez [33]byte
+	copy(rez[:], s.ID[:])
+	return rez
+}
+
+func (s ID) ToByte() []byte {
+	return s.ID[:]
+}
+
+func (s ID) IsEmpty() bool {
+	var z [33]byte
+	return bytes.Equal(s.ID[:], z[:])
+}
+
+func (s ID) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + s.String() + `"`), nil
+}
+
+type EdgeID struct {
+	ID string
+	//edge wire.OutPoint
+}
+
+//type Edge wire.OutPoint
+// Need some additional fields maybe?
+
+var NilEdgeID = EdgeID{""}
+
+func NewEdgeID(id string) EdgeID {
+	return EdgeID{id}
+}
+
+func (e EdgeID) String() string {
+	return e.ID
+}
+
+type WeighterEdge struct {
+	wgt float64
+}
+
+func NewWeighterEdge(wgt float64) *WeighterEdge {
+	return &WeighterEdge{wgt: wgt}
+}
+
+func (e *WeighterEdge) Weight() float64 {
+	return e.wgt
+}
+
+type ChannelInfo struct {
+	Cpt int64
+	Wgt float64
+}
+
+func (channel *ChannelInfo) Capacity() int64 {
+	return channel.Cpt
+}
+
+func (channel *ChannelInfo) Weight() float64 {
+	return channel.Wgt
+}
+
+func (c *ChannelInfo) Copy() *ChannelInfo {
+	if c == nil {
+		return nil
+	} else {
+		c1 := *c
+		return &c1
+	}
+}
+
+// renamed edge to allow marshalling using gob
+type Edge struct {
+	Src, Tgt ID
+	Id       EdgeID
+	InfoData *ChannelInfo
+}
+
+var NilEdge = Edge{NilID, NilID, NilEdgeID, nil}
+
+func (e Edge) String() string {
+	return fmt.Sprintf("edge[%v %v %v %v]", e.Src, e.Tgt, e.Id, e.Info)
+}
+
+func NewEdge(src, tgt ID, id EdgeID, info *ChannelInfo) Edge {
+	return Edge{
+		Src:      src,
+		Tgt:      tgt,
+		Id:       id,
+		InfoData: info,
+	}
+}
+
+func (e *Edge) Source() ID {
+	return e.Src
+}
+
+func (e *Edge) Target() ID {
+	return e.Tgt
+}
+
+func (e *Edge) ID() EdgeID {
+	return e.Id
+}
+
+func (e *Edge) Info() *ChannelInfo {
+	return e.InfoData
+}
+
+type Graph struct {
+	AdjacencyList map[ID]map[ID]map[EdgeID]*ChannelInfo
```

Comment:
The current representation allows easy determination if two vertexes have edge and allow multiple edges between vertexes(multigraph). It allows for example simple determination of all neighbors, all edges to a given neighbor. It is possible to use map[ID][]*Edge but the processing code will be much more complex(and less performant, e.g. O(number edges for node) instead of O(1) for neighborhood test).